### PR TITLE
compute-client: add protocol documentation

### LINKED ISSUE #15192: Document the compute command protocol
Recently we had to fix a number of bugs in the compute communication layer that were generally caused by a lack of understanding for how the compute protocol is supposed to work. A major hindrance is that this protocol currently is only sparsely documented in the code.

We should write a spec for the compute command protocol, that details:
* Which commands can be sent, in which order.
* What responses clients are allowed/required/forbidden to send after receiving a given command or sending a given response.

Once we have that, we will be better able to evaluate the implementation of our client layer for correctness.

### PR DESCRIPTION
**Pull request by teskje**
This commit adds documentation for the compute protocol in the form of doc strings on the `protocol` module, as well as the `ComputeCommand` and `ComputeResponse` enums.

The text is copied from [Notion](https://www.notion.so/materialize/Compute-Command-Protocol-8e1a461025d24a829db001906b0205c1), with adjustments for the doc string markdown format and updates for the things that changed since the Notion version was written (mostly `UpdateConfiguration`).

### Motivation

* This PR adds documentation.

Closes #15192.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a
  companion cloud PR to account for those changes that is tagged with
  the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

File: src/compute-client/src/protocol/command.rs [hunk 1]
```diff
@@ -34,72 +34,196 @@ include!(concat!(
...
+    /// It instructs the replica to initialize the timely dataflow runtime using the given
+    /// `config`.
+    ///
+    /// This command is special in that it is the only one that is broadcast to all processes of a
+    /// multi-process replica. All subsequent commands are only sent to the first process, which
+    /// then distributes them to the other processes using a dataflow. This method of command
+    /// distribution requires the timely dataflow runtime to be initialized, which is why the
+    /// `CreateTimely` command exists.
+    ///
+    /// The `epoch` value that imposes an ordering on iterations of the compute protocol. When the
```
Comment (vmarcos) [inline src/compute-client/src/protocol/command.rs:55]:
```suggestion
    /// The `epoch` value imposes an ordering on iterations of the compute protocol. When the
```

File: src/compute-client/src/protocol/command.rs [hunk 2]
```diff
@@ -34,72 +34,196 @@ include!(concat!(
...
     InitializationComplete,
 
-    /// Update compute instance configuration.
+    /// `UpdateConfiguration` instructs the replica to update its configuration, according to the
+    /// given [`ComputeParameter`]s.
+    ///
+    /// Parameter updates transmitted through this command must be applied by the replica as soon
+    /// as it receives the command, and they must be apply globally to all replica state, even
+    /// dataflows and pending peeks that were created before the parameter update. This property
+    /// allows the replica to hoist `UpdateConiguration` commands during reconciliation.
```
Comment (vmarcos) [inline src/compute-client/src/protocol/command.rs:92]:
```suggestion
    /// allows the replica to hoist `UpdateConfiguration` commands during reconciliation.
```

File: src/compute-client/src/protocol/response.rs [hunk 3]
```diff
@@ -27,17 +27,100 @@ include!(concat!(
...
+    ///     its read frontier to the empty frontier. ([#16275])
     ///
-    /// TODO(teskje): Consider also reporting the previous upper frontier and using that
-    /// information to assert the correct implementation of our protocols at various places.
+    /// Once a collection was reported to have been advanced to the empty upper frontier:
+    ///
+    ///   * It must no longer read from its inputs.
+    ///   * The replica must not send further `FrontierUppers` responses for that collection.
+    ///
+    /// The replica must not send send `FrontierUppers` responses for collections that have not
```
Comment (vmarcos) [inline src/compute-client/src/protocol/response.rs:63]:
```suggestion
    /// The replica must not send `FrontierUppers` responses for collections that have not
```

File: src/compute-client/src/protocol/response.rs [hunk 4]
```diff
@@ -27,17 +27,100 @@ include!(concat!(
...
+    /// The replica must not send a [`DroppedAt`] response if the subscribe’s upper frontier
+    /// (reported by [`Batch`] responses) has advanced to the empty frontier (e.g. because its
+    /// inputs advanced to the empty frontier).
+    ///
+    /// Once a subscribe was reported to have advanced to the empty frontier, or has been dropped:
+    ///
+    ///   * It must no longer read from its inputs.
+    ///   * The replica must not send further `SubscribeResponse`s for that subscribe.
+    ///
+    /// The replica must not send send [`Batch`] responses for subscribes that have not been
```
Comment (vmarcos) [inline src/compute-client/src/protocol/response.rs:114]:
```suggestion
    /// The replica must not send [`Batch`] responses for subscribes that have not been
```

**Review by vmarcos — APPROVED** [review]
Thanks, this is great work! Only minor suggestions to fix a couple of typos below.

**Comment by teskje** [comment]
TFTR!

**Pull request merged**